homeassistant/components/bmw_connected_drive/sensor.py

```diff
@@ -521,66 +173,45 @@ def __init__(
-                        date_str,
+            sensor_value = getattr(vehicle_state, sensor_key)
+
+            if isinstance(sensor_value, tuple):
+                sensor_unit = UNIT_MAP.get(sensor_value[1], sensor_value[1])
+                sensor_value = sensor_value[0]
+                sensor_value_org = sensor_value
+
+                if self.unit_of_measurement in [LENGTH_KILOMETERS, LENGTH_MILES]:
+                    sensor_value = round(
+                        self.hass.config.units.length(sensor_value, sensor_unit)
                     )
-                    self._attr_native_value = None
-            else:
-                self._attr_native_value = getattr(vehicle_last_trip, sensor_key)
-        elif self._service == SERVICE_ALL_TRIPS:
-            vehicle_all_trips = self._vehicle.state.all_trips
-            for attribute in (
-                "average_combined_consumption",
-                "average_electric_consumption",
-                "average_recuperation",
-                "chargecycle_range",
-                "total_electric_distance",
-            ):
-                if sensor_key.startswith(f"{attribute}_"):
-                    attr = getattr(vehicle_all_trips, attribute)
-                    sub_attr = sensor_key.replace(f"{attribute}_", "")
-                    self._attr_native_value = getattr(attr, sub_attr)
-                    return
-            if sensor_key == "reset_date_utc":
-                date_str = getattr(vehicle_all_trips, "reset_date")
-                if parsed_date := dt_util.parse_datetime(date_str):
-                    self._attr_native_value = parsed_date.isoformat()
-                else:
-                    _LOGGER.debug(
-                        "Could not parse date string for 'reset_date_utc' sensor: %s",
-                        date_str,
+                elif self.unit_of_measurement in [VOLUME_LITERS, VOLUME_GALLONS]:
+                    sensor_value = round(
+                        self.hass.config.units.volume(sensor_value, sensor_unit)
                     )
-                    self._attr_native_value = None
-            else:
-                self._attr_native_value = getattr(vehicle_all_trips, sensor_key)
 
-        vehicle_state = self._vehicle.state
-        charging_state = vehicle_state.charging_status in {ChargingState.CHARGING}
+                _LOGGER.debug(
+                    "UoM Conversion for %s. HA: %s %s, Vehicle: %s %s",
+                    sensor_key,
+                    sensor_value,
+                    self.unit_of_measurement,
+                    sensor_value_org,
+                    sensor_unit,
+                )
+
+        self._attr_native_value = sensor_value
 
         if sensor_key == "charging_level_hv":
+            charging_state = self._vehicle.status.charging_status in {
+                ChargingState.CHARGING
+            }
             self._attr_icon = icon_for_battery_level(
```

Comment:
Side note: The battery device class sensor already sets icon depending on charge level so we should use device class battery and remove the use of the `icon_for_battery_level` function.

https://developers.home-assistant.io/docs/core/entity/sensor#available-device-classes

Charging state should be shown by a binary sensor instead with device class `battery_charging`.

https://developers.home-assistant.io/docs/core/entity/binary-sensor#available-device-classes

Reply:
Setting the icon attribute was removed. We already have a `charging_state` as binary sensor.